### PR TITLE
Feat/tx history 02

### LINKED ISSUE #3220: Add a local cache to tx history

### PR DESCRIPTION
**Pull request by javierbitpay**
fixes #3220

### CODE REVIEW

**Comment by matiu** [comment]
👏👏👏👏

File: src/js/controllers/index.js [hunk 1]
```diff
@@ -686,20 +641,20 @@ angular.module('copayApp.controllers').controller('indexController', function($r
     var step = 6;
 
     function getHistory(skip, cb) {
-      skip = skip || 0;
-      fc.getTxHistory({
-        skip: skip,
-        limit: step,
-      }, function(err, txs) {
+      storageService.getTxHistory(c.walletId, function(err, txs) {
```
Comment (isocolsky) [inline src/js/controllers/index.js]: Where is `c` defined?

Comment (gabrielbazan7) [inline src/js/controllers/index.js]: https://github.com/javierbitpay/copay/blob/feat/tx-history-02/src/js/controllers/index.js#L665

File: src/js/controllers/index.js [hunk 2]
```diff
@@ -693,29 +649,22 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
-          });
-          return getHistory(skip + step, cb);
-        } else {
-          return cb(null, lodash.flatten(allTxs));
+
+        var txsFromLocal;
+        try {
+          txsFromLocal = JSON.parse(txs);
+        } catch (ex) {
+          return cb(ex);
```
Comment (matiu) [inline src/js/controllers/index.js]: Should not return, only $log.warn(ex);

File: src/js/controllers/index.js [hunk 3]
```diff
@@ -693,29 +649,22 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
-              console.log("Got:" + lodash.keys(unique).length + " txs");
-            } else {
-              console.log("Ignoring duplicate TX in CSV: "+ tx.txid);
-            }
-          });
-          return getHistory(skip + step, cb);
-        } else {
-          return cb(null, lodash.flatten(allTxs));
+
+        var txsFromLocal;
```
Comment (matiu) [inline src/js/controllers/index.js]: should be initialized to `[]`

File: src/js/controllers/index.js [hunk 4]
```diff
@@ -693,29 +649,22 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+          return cb(ex);
         }
+
+        allTxs.push(txsFromLocal);
+        return cb(null, lodash.flatten(allTxs));
       });
-    };
+    }
 
     if (isCordova) {
```
Comment (matiu) [inline src/js/controllers/index.js]: This check should not be performed, this function should never be called if `!isCordoba`.

Or it should throw:
if (isCordoba) throw ('CSV generation not available in mobile');

File: src/js/controllers/index.js [hunk 5]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
   };
 
+  self.stopSync = function(remoteTx, localTx) {
+    if (remoteTx.txid == localTx.txid)
+      return true;
+    else
+      return false;
+  }
+
+  self.removeLessThanSixConfirmations = function(txs) {
```
Comment (matiu) [inline src/js/controllers/index.js]: should be `removeSoftConfirmedTx`

File: src/js/controllers/index.js [hunk 6]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+    return lodash.map(txs, function(tx) {
+      if (tx.confirmations >= CONFIRMATIONS)
```
Comment (matiu) [inline src/js/controllers/index.js]: Constant name should be SOFT_CONFIRMATION_LIMIT

File: src/js/controllers/index.js [hunk 7]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+      } catch (ex) {
+        return cb(ex);
+      }
+
+      if (!localTxs)
+        localTxs = [];
+
+      var txsFromLocal = self.removeLessThanSixConfirmations(localTxs);
+
+      var count = 0;
```
Comment (matiu) [inline src/js/controllers/index.js]: remove unused var

File: src/js/controllers/index.js [hunk 8]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+        return tx;
+    });
+  }
+
+  self.updateLocalTxHistory = function(cb) {
+    var fc = profileService.focusedClient;
+    var c = fc.credentials;
+    var txsToPush = [];
+
+    storageService.getTxHistory(c.walletId, function(err, txs) {
```
Comment (matiu) [inline src/js/controllers/index.js]: it seems that from 757 to 771 could be a new function:
`self.getLocalTxs(function(err, localTxs) {`

File: src/js/controllers/index.js [hunk 9]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+    return lodash.map(txs, function(tx) {
+      if (tx.confirmations >= CONFIRMATIONS)
+        return tx;
+    });
+  }
+
+  self.updateLocalTxHistory = function(cb) {
+    var fc = profileService.focusedClient;
+    var c = fc.credentials;
+    var txsToPush = [];
```
Comment (matiu) [inline src/js/controllers/index.js]: remove unused var `txsToPush`

File: src/js/controllers/index.js [hunk 10]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+        return cb(ex);
+      }
+
+      if (!localTxs)
+        localTxs = [];
+
+      var txsFromLocal = self.removeLessThanSixConfirmations(localTxs);
+
+      var count = 0;
+      fillTxsObject(txsToPush);
```
Comment (matiu) [inline src/js/controllers/index.js]: pass ([])

File: src/js/controllers/index.js [hunk 11]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+
+      function fillTxsObject(txsToPush) {
+        self.makeTxHistoryRequest(txsToPush, txsFromLocal, function(err, exitLoop, txsResult) {
+          if (err) return cb(err);
+          if (exitLoop) {
+            self.txHistory = [];
```
Comment (matiu) [inline src/js/controllers/index.js]: Good job here!

File: src/js/controllers/index.js [hunk 12]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+      return cb(null, exitLoop, txsToPush);
+    });
+  }
+
+  self.updateHistory = function() {
+    $log.debug('Updating Transaction History');
+    self.skipHistory = 0;
+    self.txHistoryError = false;
+    self.updatingTxHistory = true;
+    self.txHistoryPaging = false;
```
Comment (matiu) [inline src/js/controllers/index.js]: rm unused var `txHistoryPaging`? also from HTML?

File: src/js/controllers/index.js [hunk 13]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+        }
+      });
+      self.skipHistory = self.skipHistory + self.limitHistory;
+      return cb(null, exitLoop, txsToPush);
+    });
+  }
+
+  self.updateHistory = function() {
+    $log.debug('Updating Transaction History');
+    self.skipHistory = 0;
```
Comment (matiu) [inline src/js/controllers/index.js]: rm unused var skipHistory? also from HTML?

File: src/js/controllers/index.js [hunk 14]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+    });
+  }
+
+  self.makeTxHistoryRequest = function(txsToPush, localTx, cb) {
+    var fc = profileService.focusedClient;
+    var c = fc.credentials;
+    var exitLoop = false;
+
+    fc.getTxHistory({
+      skip: self.skipHistory,
```
Comment (matiu) [inline src/js/controllers/index.js]: it seems that skipHistory should be a param in `makeTxHistoryRequest`, and not a "global" variable.

File: src/js/controllers/index.js [hunk 15]
```diff
@@ -782,6 +735,114 @@ angular.module('copayApp.controllers').controller('indexController', function($r
...
+              return cb(null);
+            });
+          } else
+            fillTxsObject(txsResult);
+        });
+      };
+    });
+  }
+
+  self.makeTxHistoryRequest = function(txsToPush, localTx, cb) {
```
Comment (matiu) [inline src/js/controllers/index.js]: instead of passing `localTx[]`, it should use `lastTx` (only one TX, not an array).

**Comment by javierbitpay** [comment]
Updated.

**Comment by matiu** [comment]
Great job!

**Pull request merged**